File: my-folder/0309-best-time-to-buy-and-sell-stock-with-cooldown/solution.go

func maxProfit(prices []int) int {
    n := len(prices)
    dp := make([][]int, n+1)

    for i := range dp {
        dp[i] = []int{-1,-1}
    }
    
    var dfs func(int, int) int
    dfs = func(i int, isBuy int) int {
        if i >= len(prices) {
            return 0
        }

        if dp[i][isBuy] != -1 {
            return dp[i][isBuy]
        }

        cooldown := dfs(i+1, isBuy)

        if isBuy == 1 {
            buy := dfs(i+1, 0) - prices[i]
            dp[i][isBuy] = max(buy, cooldown)
        } else {
            sell := dfs(i+2, 1) + prices[i]
            dp[i][isBuy] = max(sell, cooldown)
        }

        return dp[i][isBuy]
    }

    return dfs(0, 1)
}

File: my-folder/0309-best-time-to-buy-and-sell-stock-with-cooldown/solution.rs
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut hold = -prices[0];
        let mut sold = 0;
        let mut rest = 0;

        for i in 0..prices.len() {
            let new_hold = hold.max(rest - prices[i]);
            let new_sold = hold + prices[i];
            let new_rest = rest.max(sold);

            hold = new_hold;
            sold = new_sold;
            rest = new_rest;
        }

        return rest.max(sold);
    }
}
